src/coreclr/jit/lclmorph.cpp

```diff
@@ -1317,7 +1317,13 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 
         if (addr->IsLclVarAddr())
         {
-            const LclVarDsc* varDsc = m_compiler->lvaGetDesc(addr->AsLclVarCommon());
+            LclVarDsc* varDsc = m_compiler->lvaGetDesc(addr->AsLclVarCommon());
+
+            // addr is a span field so this lclVar must be a span
+            if (isSpanLength)
+            {
+                varDsc->SetIsSpan(true);
+            }
```

Comment:
With the handling in `lvaSetStruct` I wouldn't expect this to be necessary.

Reply:
OTOH, I think it would be beneficial to do this in the other direction inside `PromoteStructVar`. When we create the local for the length field there we can mark it as never negative for spans.

Reply:
Afaik optimization in this PR only kicked in when span was unpromoted. At least that's what I heard before.

Reply:
The optimization you added in `IntegralRange` only kicks in then, but we no longer need to mark as `varDsc` as a span when we see the length intrinsic -- the code you added in `lvaSetStruct` handles that much more generally.